File: src/controllers/cart.controller.ts
import { NextFunction, Request, Response } from 'express';
import { CartService } from '../services/cart.service';
import { UnauthorizedError } from '../errors/AppError';

function requireUserId(req: Request): string {
  if (!req.user) {
    throw new UnauthorizedError();
  }
  return req.user.id;
}

export class CartController {
  constructor(private readonly cartService: CartService) {}

  getCart = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const cart = await this.cartService.getCart(requireUserId(req));
      res.status(200).json(cart);
    } catch (error) {
      next(error);
    }
  };

  addItem = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userId = requireUserId(req);
      const cart = await this.cartService.addItem(userId, req.body.productId, req.body.quantity);
      res.status(201).json(cart);
    } catch (error) {
      next(error);
    }
  };

  updateItem = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userId = requireUserId(req);
      const cart = await this.cartService.updateItemQuantity(userId, req.params.id, req.body.quantity);
      res.status(200).json(cart);
    } catch (error) {
      next(error);
    }
  };

  removeItem = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userId = requireUserId(req);
      await this.cartService.removeItem(userId, req.params.id);
      res.status(204).send();
    } catch (error) {
      next(error);
    }
  };
}

File: src/Controllers/CartController.cs
using ECommerceApi.DTOs.Cart;
using ECommerceApi.Extensions;
using ECommerceApi.Services.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ECommerceApi.Controllers;

// Every action requires authentication
[ApiController]
[Route("api/cart")]
[Authorize]
public class CartController : ControllerBase
{
    private readonly ICartService _cartService;

    public CartController(ICartService cartService)
    {
        _cartService = cartService;
    }

    [HttpGet]
    public async Task<IActionResult> GetCart()
    {
        var cart = await _cartService.GetCartAsync(this.GetUserId());
        return Ok(cart);
    }

    [HttpPost("items")]
    [ProducesResponseType(typeof(CartResponse), StatusCodes.Status201Created)]
    public async Task<IActionResult> AddItem([FromBody] AddCartItemRequest request)
    {
        var cart = await _cartService.AddItemAsync(this.GetUserId(), request.ProductId, request.Quantity);
        return StatusCode(StatusCodes.Status201Created, cart);
    }

    [HttpPut("items/{productId:guid}")]
    public async Task<IActionResult> UpdateItem(Guid productId, [FromBody] UpdateCartItemRequest request)
    {
        var cart = await _cartService.UpdateItemQuantityAsync(this.GetUserId(), productId, request.Quantity);
        return Ok(cart);
    }

    [HttpDelete("items/{productId:guid}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> RemoveItem(Guid productId)
    {
        await _cartService.RemoveItemAsync(this.GetUserId(), productId);
        return NoContent();
    }
}
